pkg/cli/admin/ocpcertificates/regeneratemco/command.go

```diff
@@ -0,0 +1,297 @@
+		clientset,
+		24*time.Hour,
+		informers.WithNamespace(o.mcoNamespace))
+
+	caName := mcsName + "-ca"
+	cont := certrotation.NewCertRotationController(
+		controllerName,
+		certrotation.RotatedSigningCASecret{
+			Namespace:     o.mcoNamespace,
+			Name:          caName,
+			Validity:      caExpiry,
+			Refresh:       caRefresh,
+			Informer:      inf.Core().V1().Secrets(),
+			Lister:        inf.Core().V1().Secrets().Lister(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: recorder,
+		},
+		certrotation.CABundleConfigMap{
+			Namespace:     o.mcoNamespace,
+			Name:          caName,
+			Lister:        inf.Core().V1().ConfigMaps().Lister(),
+			Informer:      inf.Core().V1().ConfigMaps(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: recorder,
+		},
+		certrotation.RotatedSelfSignedCertKeySecret{
+			Namespace: o.mcoNamespace,
+			Name:      mcsTlsSecretName,
+			Validity:  keyExpiry,
+			Refresh:   keyRefresh,
+			CertCreator: &certrotation.ServingRotation{
+				Hostnames: func() []string { return []string{apiserverIntURL.Hostname()} },
+			},
+			Lister:        inf.Core().V1().Secrets().Lister(),
+			Informer:      inf.Core().V1().Secrets(),
+			Client:        clientset.CoreV1(),
+			EventRecorder: recorder,
+		},
+		nil, // no operatorclient needed
+		recorder,
+	)
+
+	ch := make(chan struct{})
+	inf.Start(ch)
+	inf.WaitForCacheSync(ch)
+
+	ctx := context.WithValue(context.Background(), certrotation.RunOnceContextKey, true)
+	syncCtx := factory.NewSyncContext(mcsName, recorder)
+	if err := cont.Sync(ctx, syncCtx); err != nil {
+		return err
+	}
+
+	fmt.Println("Successfully rotated MCS CA + certs.")
+
+	// Redeploy MCS. This will eventually not be needed, see: https://github.com/openshift/machine-config-operator/pull/3744
+	mcoPods := clientset.CoreV1().Pods(mcoNamespace)
+	mcsPods, err := mcoPods.List(ctx, metav1.ListOptions{
+		LabelSelector: mcsLabelSelector,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot get MCS pods: %w", err)
+	}
+	for _, pod := range mcsPods.Items {
+		err := mcoPods.Delete(ctx, pod.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot delete MCS pod %s: %w", pod.Name, err)
+		}
+	}
+
+	// TODO maybe add a watcher to make sure the MCS daemonset is ready here
+
+	// Get the updated CA cert
+	mcoSecrets := clientset.CoreV1().Secrets(mcoNamespace)
+	mcsSecret, err := mcoSecrets.Get(ctx, newMCSCASecret, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot read MCS secret: %w", err)
+	}
+
+	caCert, exists := mcsSecret.Data[CASecretCertName]
+	if !exists {
+		return fmt.Errorf("could not find MCS CA cert at %s after rotation", CASecretCertName)
+	}
+	encodedCACert := base64.StdEncoding.EncodeToString(caCert)
+
+	if o.outputNewCert {
+		fmt.Printf("New Certificate:\n%s\n", encodedCACert)
+	}
+
+	if !o.modifyUserData {
+		return nil
+	}
+
+	// Get user-data-secret
+	mapiSecrets := clientset.CoreV1().Secrets(mapiNamespace)
```

Comment:
Based on usage and potential failures, I think everything from here down is logically a different command.

Reply:
Hmm.  Yes, I see your point.  On the other hand:

- these secret objects are admin owned and not expected to change by anything else
- The cert code is already a "run once" controller and hence is idempotent and can be run multiple times reliably
- We really do want to update both of these things in "one go" as our default stance; doing one but not the other will result in a broken state

Reply:
@yuqi-zhang I'm ok with that approach.